state_synchronizer/src/coordinator.rs

```diff
@@ -205,37 +219,51 @@ impl<T: ExecutorProxyTrait> SyncCoordinator<T> {
     }
 
     /// Get a batch of transactions
-    async fn process_chunk_request(&mut self, peer_id: PeerId, mut request: GetChunkRequest) {
-        let latest_ledger_info = match self.executor_proxy.get_latest_ledger_info().await {
-            Ok(li) => li,
-            Err(err) => {
-                error!("[state sync] failed to fetch latest ledger info: {:?}", err);
-                return;
-            }
-        };
+    async fn process_chunk_request(
+        &mut self,
+        peer_id: PeerId,
+        mut request: GetChunkRequest,
+    ) -> Result<()> {
+        let latest_ledger_info = self.executor_proxy.get_latest_ledger_info().await?;
         let target = LedgerInfo::from_proto(request.take_ledger_info_with_sigs())
             .unwrap_or(latest_ledger_info);
 
-        match self
-            .executor_proxy
-            .get_chunk(request.known_version, request.limit, target)
-            .await
-        {
-            Ok(response) => {
-                let mut msg = StateSynchronizerMsg::new();
-                msg.set_chunk_response(response);
-                if self.network_sender.send_to(peer_id, msg).await.is_err() {
-                    error!("[state sync] failed to send p2p message");
-                }
-            }
-            Err(err) => {
-                error!("[state sync] executor error {:?}", err);
+        // if upstream synchronizer doesn't have new data and request timeout is set
+        // add peer request into subscription queue
+        if self.known_version <= request.known_version && request.timeout > 0 {
+            let expiration_time =
+                SystemTime::now().checked_add(Duration::from_millis(request.timeout));
+            if let Some(time) = expiration_time {
+                self.subscriptions
+                    .insert(peer_id, (time, request.known_version, request.limit));
             }
+            return self.check_subscriptions().await;
```

Comment:
should we check subscriptions periodically as compared to doing so on `process_chunk_request` calls?

Reply:
we don't need to do it. Subscription can be ready only if there's new data. New data (consensus commit block on Validator or downstream push on FullNode) would trigger `commit` function call. Commit handler inside checks ready subscriptions